doc/image_processing/contrast_enhancement/histogram_equalization.rst

```diff
@@ -0,0 +1,100 @@
+.. _he:
+
+######################
+Histogram Equalization 
+######################
+
+Description
+-----------
+
+Histogram equalization also known as histogram flattening, is a non-linear image enhancement
+algorithm that follows the idea that not only should an image cover the entire grayscale space
+but also be uniformly distributed over that range. 
+
+An ideal image would be the one having a flat histogram.
+
+Although care should be taken before applying a non-linear transformation on the image
+histogram, there are good mathematical reasons why a flat histogram is the desired goal.
+
+A simple scenario would be an image with pixels concentrated in an interval, in which case
+histogram equalization transforms pixels to achieve a flat histogram image. Thus enhancing
+the image contrast.
+
+.. figure:: he_chart.png
+    :width: 200px
+    :align: center
+    :height: 100px
+    :alt: Could not load image.
+    :figclass: align-center
+
+    Pixels concentrated in an interval spread out.
+
+Algorithm
+---------
+
+#. First calculate the histogram corresponding to input image.
+#. If it is a multi channeled image (e.g. RGB), convert it to a independent color space
+   (like YCbCr, HSV etc.).
+#. Then calculate the cumulative histogram over the input image.
+#. Normalize the histogram to bring bin values between 0-1. For multi-channeled images 
+   normalize each channel independently (by the number of pixels in image).
+#. If the histogram of image is H(p\ :sub:`x`\)  p\ :sub:`x`\  in  [0, 255], then apply 
+   the transformation p\ :sub:`x'`\  = H(p\ :sub:`x`\),  p\ :sub:`x'`\  is pixel in output
+   image.
+
+**Explanation**
+
+Since we will be transforming the image to match a flat histogram, we match
+the cumulative histogram of the image to the cumulative histogram of a flat histogram.
+
+Cumulative histogram of flat image is  H(p\ :sub:`x'`\)  =  p\ :sub:`x'` .
+
+Hence,
+
+        =>  H(p\ :sub:`x'`\)  =  H(p\ :sub:`x`\)
+
+        =>  p\ :sub:`x'`\     =  H(p\ :sub:`x`\)
+
+Results
+-------
+The algorithm is applied on a few standard images. One of the transformations in shown below:
+
+**Grayscale Image**
+
+.. figure:: barbara.png
+    :width: 600px
+    :align: center
+    :height: 300px
+    :alt: Could not load image.
+    :figclass: align-center
+
+**RGB**
+
+.. figure:: church.png
+    :width: 600px
+    :align: center
+    :height: 250px
+    :alt: Could not load image.
+    :figclass: align-center
```

Comment:
I do not like the idea of storing 7MB files in GIL's repo.
Unfortunately, I can not offer an ideal alternative yet.

I'd suggest a temporary solution:
1. Remove the images from this PR
2. Paste the images as comments to this PR
3. Replace filename with URLs from GitHub

```
.. figure:: https://user-images.githubusercontent.com/12345/12811123-3877a980-8375-11ea-9960-1b88a3f9f87b.png
...
```

The thing is, once we start adding big binary files to Git repo, then even if we find an optimal solution for hosting images for docs, we will not be able to get rid of them easily, from Git history.

I have (slowly) started collecting externally hosted test images in dedicated repo https://github.com/boost-gil/test-images, so I perhaps we could have `examples/` folder there and link from docs to there.

What do you think @lpranam ?

Reply:
In my opinion, in docs we don't have to use the full-size images. Docs just need to demonstrate what our algorithms can do and do not need to show the exact accurate images, we can use highly compressed and small resolution images in docs. Otherwise what @mloskot suggested the great. I have no other objections regarding this PR.

Reply:
> @codejaeger Well done! It's great to see GSoC features covered with documentation :-) I'm happy to approve this PR.
> 

Thanks a lot @mloskot : )  

> I'd suggest we merge it **after** we merge PR #499.
> Next time you may prefer to stick docs in the same PR as implementation and tests.

Yes sure. Also this PR contains docs for the algorithm histogram equalization. I haven't pushed it yet since the histogram class is a dependency for it. Should I remove the docs for the algorithm?

Reply:
I have compressed the images, they now occupy about 34 KB and 24 KB. I can do what @mloskot suggested if needed.

Reply:
though it is compressed for doc would be great to add real one on that repo so we can have those images for later.

Reply:
@codejaeger If you want to try the idea outlined earlier, you should have now the write access to https://github.com/boost-gil/test-images Feel free to create `doc/` directory (and any necessary subfolders) and put your images in there.

Reply:
@codejaeger Could you add full size/resolution images to the `test-images` repo as @lpranam suggested?

Reply:
@codejaeger Ping ☝️

Reply:
Yep doing it. Thanks for the reminder, I had totally forgotten about it.